woocommerce/google-listings-and-ads
# Upgrade Google Ads library to V11

### Changes proposed in this Pull Request:

This PR updates the Google Ads library from V9 > V11. Since that requires various dependencies to be updated, here is an outline of all the things that were changed:

- Bump minimum PHP version to 7.4
- Remove unit testing for PHP 7.3
- Bump WP versions in unit tests to be L-2
- Update dependent composer packages including Guzzle, replaces PR #1571 
- Update the Google Client services (client package was already updated in #1559)
- Change all imports from V9 > V11
- Cleanup older versions of the Ads library V8 - V10
- Update cleanup script to support V11
- One of the helper functions for parsing the resourceName was removed (this is reported [here](https://github.com/googleapis/gapic-generator-php/issues/453), since they aren't available anymore for all levels), this PR replaces the helper function with a regular expression to parse the ID
- GET functions were removed for several services (recommended is to use a search query instead), in one of the tests the exception was still mocking the GET request, so this was changed to mock a query exception
- Along with the previous point the ConversionAction is now also retrieved through a search query (tests were updated to match)

### Updated packages
```
composer update php
composer update googleads/google-ads-php --with-all-dependencies

- Upgrading firebase/php-jwt (v6.2.0 => v6.3.0)
- Upgrading google/auth (v1.21.0 => v1.21.1)
- Upgrading google/gax (v1.12.2 => v1.13.0)
- Upgrading google/protobuf (v3.21.1 => v3.21.2)
- Upgrading googleads/google-ads-php (v12.1.0 => v16.0.0)
- Upgrading guzzlehttp/guzzle (7.4.4 => 7.4.5)
- Upgrading guzzlehttp/psr7 (2.3.0 => 2.4.0)
- Upgrading symfony/deprecation-contracts (v2.5.1 => v2.5.2)

composer update google/apiclient --with-all-dependencies

- Upgrading google/apiclient-services (v0.252.0 => v0.258.0)
- Upgrading paragonie/constant_time_encoding (v2.5.0 => v2.6.3)
```

Closes #1600

### Detailed test instructions:
1. Checkout this PR, remove the vendors folder and run `composer install` 
2. Run the unit tests locally to confirm they all pass without errors or warnings
3. In the Settings page disconnect any previously connected Ads Account
4. Onboard the same Ads account again
5. Confirm that all the steps complete without issues
6. Check the option `gla_ads_conversion_action` in the `wp_options DB table, and make sure it contains something like:
```
a:5:{
    s:2:"id";i:987654321;
    s:4:"name";s:46:"[a12b] Google Listings and Ads purchase action";
    s:6:"status";s:7:"ENABLED";
    s:13:"conversion_id";s:12:"AW-12345678";
    s:16:"conversion_label";s:20:"abc_DEF";
}
```
7. View the list of campaigns on the Dashboard page including some report data to confirm that there are no errors when using the update library


### Additional details:
As a follow up to this issue we can also update the WCS to use API V11,  PR 1883-gh-Automattic/woocommerce-connect-server can be used as a template for what needs to be changed.

### Changelog entry
* Update - Google Ads library to API V11.
